server/src/services/semanticHighlight/onSemanticTokensFull.ts

```diff
@@ -0,0 +1,159 @@
+import {
+  SemanticTokens,
+  SemanticTokensParams,
+} from "vscode-languageserver-protocol";
+import _, { Dictionary } from "lodash";
+import { analyze } from "@nomicfoundation/solidity-analyzer";
+import semver from "semver";
+import { Language } from "@nomicfoundation/slang/language";
+import { ProductionKind, TokenKind } from "@nomicfoundation/slang/kinds";
+import { Cursor } from "@nomicfoundation/slang/cursor";
+import { TokenNode } from "@nomicfoundation/slang/cst";
+import { ServerState } from "../../types";
+import { SlangNode } from "../../parser/slangHelpers";
+import { CustomTypeHighlighter } from "./highlighters/CustomTypeHighlighter";
+import { SemanticTokensBuilder } from "./SemanticTokensBuilder";
+import { FunctionDefinitionHighlighter } from "./highlighters/FunctionDefinitionHighlighter";
+import { FunctionCallHighlighter } from "./highlighters/FunctionCallHighlighter";
+import { EventEmissionHighlighter } from "./highlighters/EventEmissionHighlighter";
+import { EventDefinitionHighlighter } from "./highlighters/EventDefinitionHighlighter";
+import { ContractDefinitionHighlighter } from "./highlighters/ContractDefinitionHighlighter";
+import { InterfaceDefinitionHighlighter } from "./highlighters/InterfaceDefinitionHighlighter";
+import { StructDefinitionHighlighter } from "./highlighters/StructDefinitionHighlighter";
+import { HighlightVisitor } from "./HighlightVisitor";
+import { UserDefinedValueTypeDefinitionHighlighter } from "./highlighters/UserDefinedValueTypeDefinitionHighlighter copy";
+import { EnumDefinitionHighlighter } from "./highlighters/EnumDefinitionHighlighter";
+import { ErrorDefinitionHighlighter } from "./highlighters/ErrorDefinitionHighlighter";
+import { LibraryDefinitionHighlighter } from "./highlighters/LibraryDefinitionHighlighter";
+
+const emptyResponse: SemanticTokens = { data: [] };
+
+export function onSemanticTokensFull(serverState: ServerState) {
+  return (params: SemanticTokensParams): SemanticTokens => {
+    const { telemetry, logger } = serverState;
+
+    return (
+      telemetry.trackTimingSync("onSemanticTokensFull", (transaction) => {
+        const { uri } = params.textDocument;
+
+        // Find the file in the documents collection
+        const document = serverState.documents.get(uri);
+
+        if (document === undefined) {
+          logger.error("document not found in collection");
+          return {
+            status: "internal_error",
+            result: emptyResponse,
+          };
+        }
+
+        const text = document.getText();
+
+        // Get the document's solidity version
+        let span = transaction.startChild({ op: "solidity-analyzer" });
+        const { versionPragmas } = analyze(text);
+        span.finish();
+
+        const versions = Language.supportedVersions();
+        versionPragmas.push(
+          `>= ${versions[0]}`,
+          `<= ${versions[versions.length - 1]}`
+        );
+
+        const solcVersion = semver.maxSatisfying(
```

Comment:
left over:

```suggestion
        const slangVersion = semver.maxSatisfying(
```

Reply:
Also, I suggest reusing `versions` created above instead of fetching it two more times.